src/Take.Elephant.Kafka/PartitionedKafkaReceiver.cs

```diff
@@ -0,0 +1,177 @@
+using Confluent.Kafka;
+
+namespace Take.Elephant.Kafka
+{
+    public class PartitionedKafkaReceiver<TKey, TEvent> : IEventStreamConsumer<TKey, TEvent>, IOpenable, ICloseable, IDisposable
+    {
+        private readonly IConsumer<TKey, TEvent> _partitionedConsumer;
+        private readonly SemaphoreSlim _consumerStartSemaphore;
+        private readonly CancellationTokenSource _cts;
+        private readonly Channel<Message<TKey, TEvent>> _channel;
+        private Task _consumerTask;
+        private bool _closed;
+
+        public PartitionedKafkaReceiver(string bootstrapServers, string topic, string groupId, ISerializer<TEvent> serializer, IDeserializer<TEvent> deserializer = null)
+            : this(new ConsumerConfig() { BootstrapServers = bootstrapServers, GroupId = groupId }, topic, serializer, deserializer)
+        {
+        }
+
+        public PartitionedKafkaReceiver(
+            ConsumerConfig consumerConfig,
+            string topic,
+            ISerializer<TEvent> serializer,
+            IDeserializer<TEvent> deserializer = null)
+            : this(
+                  new ConsumerBuilder<TKey, TEvent>(consumerConfig)
+                      .SetValueDeserializer(deserializer ?? new Deserializer<TEvent>(serializer))
+                      .Build(),
+                    serializer,
+                  topic)
+        {
+        }
+
+        public PartitionedKafkaReceiver(IConsumer<TKey, TEvent> consumer, ISerializer<TEvent> serializer, string topic)
+        {
+            if (string.IsNullOrWhiteSpace(topic))
+                throw new ArgumentException("Value cannot be null or whitespace.", nameof(topic));
+            _partitionedConsumer = consumer ?? throw new ArgumentNullException(nameof(consumer));
+            Topic = topic;
+            _consumerStartSemaphore = new SemaphoreSlim(1, 1);
+            _cts = new CancellationTokenSource();
+            _channel = Channel.CreateBounded<Message<TKey, TEvent>>(1);
+        }
+
+        public string Topic { get; }
+
+        public virtual async Task<(TKey key, TEvent item)> ConsumeOrDefaultAsync(CancellationToken cancellationToken)
+        {
+            await StartConsumerTaskIfNotAsync(cancellationToken);
+            if (_channel.Reader.TryRead(out var message))
+            {
+                return (message.Key, message.Value);
+            }
+
+            return default;
+        }
+
+        public Task OpenAsync(CancellationToken cancellationToken)
+        {
+            return StartConsumerTaskIfNotAsync(cancellationToken);
+        }
+
+        public virtual Task CloseAsync(CancellationToken cancellationToken)
+        {
+            if (!_closed)
+            {
+                _partitionedConsumer.Close();
+                _closed = true;
+            }
+
+            return _consumerTask;
+        }
+
+        public event EventHandler<ExceptionEventArgs> ConsumerFailed;
+
+        protected virtual void Dispose(bool disposing)
```

Comment:
the semaphore is also `IDisposable`, thus its `Dispose` method should also be called here